server/src/weather.rs

```diff
@@ -2,86 +2,120 @@ use crate::{network::Network, Game, World};
 use feather_core::network::packet::implementation::ChangeGameState;
 use fecs::Entity;
 use rand::Rng;
+use std::cmp;
 
 const TICKS_DAY: i32 = 24_000;
 const TICKS_HALF_DAY: i32 = TICKS_DAY / 2;
 const TICKS_WEEK: i32 = TICKS_DAY * 7;
 const THUNDER_FACTOR: i32 = 10;
 
-#[derive(PartialEq)]
+#[derive(PartialEq, Clone, Copy)]
 pub enum Weather {
     Clear,
     Rain,
     Thunder,
 }
 
+#[derive(Clone, Copy)]
 pub struct WeatherChangeEvent {
     pub from: Weather,
     pub to: Weather,
+    pub duration: i32,
+}
+
+pub fn clear_weather(game: &mut Game) {
+    let durration = game
+        .rng()
+        .gen_range(TICKS_HALF_DAY, TICKS_WEEK + TICKS_HALF_DAY);
+    set_weather(game, Weather::Clear, durration);
 }
 
 #[system]
 pub fn handle_weather(game: &mut Game, world: &mut World) {
-    let from = game.weather();
-
-    if game.level.rain_time <= 0 {
-        game.level.raining = !game.level.raining;
-
-        let duration = if game.level.raining {
-            game.rng().gen_range(TICKS_HALF_DAY, TICKS_DAY)
-        } else {
-            game.rng()
-                .gen_range(TICKS_HALF_DAY, TICKS_WEEK + TICKS_HALF_DAY)
-        };
-
-        game.level.rain_time = duration;
+    if game.level.clear_weather_time >= 0 {
+        game.level.clear_weather_time -= 1;
+        return;
     }
 
-    if game.level.thunder_time <= 0 {
-        game.level.thundering = !game.level.thundering;
+    let from = game.weather();
 
-        let duration = if game.level.raining {
-            game.rng().gen_range(TICKS_HALF_DAY, TICKS_DAY)
+    game.level.rain_time -= 1;
+    let mut to = if game.level.rain_time <= 0 {
+        if game.level.raining {
+            Weather::Clear
         } else {
-            // Let the time between thunderstorms be THUNDER_FACTOR larger than normal rain.
-            // Ie. roughly 10 normal rain between thunderstorms.
-            game.rng()
-                .gen_range(TICKS_HALF_DAY, TICKS_WEEK + TICKS_HALF_DAY)
-                * THUNDER_FACTOR
-        };
-
-        game.level.thunder_time = duration;
-    }
+            Weather::Rain
+        }
+    } else {
+        from
+    };
 
-    let to = game.weather();
+    game.level.thunder_time -= 1;
+    to = if game.level.thunder_time <= 0 {
+        if game.level.thundering {
+            Weather::Clear
+        } else {
+            Weather::Thunder
+        }
+    } else {
+        to
+    };
 
     if from != to {
-        game.on_weather_change(world, WeatherChangeEvent { from, to })
+        let duration = match to {
+            Weather::Clear => game
+                .rng()
+                .gen_range(TICKS_HALF_DAY, TICKS_WEEK + TICKS_HALF_DAY),
+            _ => game.rng().gen_range(TICKS_HALF_DAY, TICKS_DAY),
+        };
+        let mut event = WeatherChangeEvent { from, to, duration };
```

Comment:
How should I handle events? Or should events only notify of changes

Reply:
This seems about right to me.